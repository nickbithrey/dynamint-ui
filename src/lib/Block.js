import React from 'react';
import { Stack } from 'office-ui-fabric-react/lib/Stack';

const Block = ({children, ...remainingProps}) => (
	<Stack.Item {...remainingProps}>
		{children}
	</Stack.Item>
);

export default Block;