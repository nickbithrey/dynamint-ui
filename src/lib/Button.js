import React from 'react';
import PropTypes from 'prop-types';
import { 
	PrimaryButton, 
	DefaultButton, 
	CommandBarButton, 
	IconButton 
} from 'office-ui-fabric-react/lib/Button';

const buttonTypes = {
	standard: DefaultButton,
	primary: PrimaryButton,
	command: CommandBarButton,
	icon: IconButton
};

const Button = ({text, onClick, type}) => {
	const Btn = buttonTypes[type];
	return <Btn text={text} onClick={onClick} />
};

Button.defaultProps = {
	type: 'standard'
};

Button.propTypes = {
	text: PropTypes.string.isRequired,
	onClick: PropTypes.func
}

export default Button;

export const newBtn = (text, onClick, condition = () => true) => (
	{
		text,
		onClick,
		condition,
		type: 'standard'
	}
);

export const newPrimaryBtn = (text, onClick, condition = () => true) => (
	{
		text,
		onClick,
		condition,
		type: 'primary'
	}
);