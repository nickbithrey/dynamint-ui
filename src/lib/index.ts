export interface Key {
    key: string | number;
}

export interface Condition<T> {
    condition: (input: T) => boolean;
}

export interface Loading {
    loading: boolean;
}

export interface Items<T> {
    items: Array<T>;
}

export interface URI {
    uri: string;
}

export interface Display {
    readOnly: boolean;
}

export interface IURILink {
    href: string;
}