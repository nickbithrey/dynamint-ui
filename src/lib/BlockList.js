import React from 'react';
import PropTypes from 'prop-types';
import { List } from 'office-ui-fabric-react/lib/List';

const renderCell = (tag, updateEntry) => {
	const CellTag = tag;
	return (item, index) => {
		return <CellTag item={item} index={index} update={updateEntry} />;
	};
};

const BlockList = ({items, tag, updateEntry}) => {
	return (
		<List
	        items={items}
	        onRenderCell={renderCell(tag, updateEntry)}
	    />
	);
};

BlockList.propTypes = {
	items: PropTypes.arrayOf(PropTypes.object).isRequired,
	tag: PropTypes.func.isRequired
}

export default BlockList;