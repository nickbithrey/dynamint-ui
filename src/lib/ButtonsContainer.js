import React from 'react';
import Btn from '~/lib/Button';
import { Columnar, Block } from '~/lib/Grid';

const ButtonsContainer = ({children = [], buttons = []}) => (
	<Columnar horizontalAlign={'end'} tokens={{childrenGap: 5}}>
		{renderChildren(children)}
		{buttons.map(renderButton)}
	</Columnar>
);

const renderChildren = (children) => {
	if (Array.isArray(children)) {
		return children.map(renderChild)
	} else if (children) {
		return renderChild(children, 0);
	}
}

const renderChild = (child, i) => (
	<Block key={i}>
		{child}
	</Block>
);

const renderButton = (button, i) => {
	if (!button.condition()) {
		return;
	}
	return (
		<Block key={i}>
			<Btn {...button} />
		</Block>
	);
}

export default ButtonsContainer;