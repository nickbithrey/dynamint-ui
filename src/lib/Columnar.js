import React from 'react';
import { Stack } from 'office-ui-fabric-react/lib/Stack';

const Columnar = ({children, ...remainingProps}) => (
	<Stack horizontal horizontalAlign="space-between" {...remainingProps}>
		{children}
	</Stack>
);

export default Columnar;