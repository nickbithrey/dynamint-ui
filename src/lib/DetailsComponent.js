import React from 'react';
import PropTypes from 'prop-types';
import Button from '~/lib/Button';
import Loading from '~/lib/Loading';

const detailsComponent = (WrappedComponent) => {
	class DetailsComponent extends React.Component {
		
		constructor(props) {
			super(props);
			this.load = this.load.bind(this);
			this.renderTitle = this.renderTitle.bind(this);
			this.renderBackBtn = this.renderBackBtn.bind(this);
			this.back = this.back.bind(this);
			this.handleUpdateField = this.handleUpdateField.bind(this);
			this.handleUpdate = this.handleUpdate.bind(this);
			this.uri = props.location.state.uri;
			this.isCreate = props.location.state.isCreate;
		}
		
		load() {
			this.props.load(this.uri, this.isCreate);
		}
		
		componentDidMount() {
			this.load();
		}
		
		componentWillUnmount() {
			this.props.clear();
		}
		
		handleUpdateField(field) {
			this.props.updateField(field);
		}
		
		handleUpdate(e) {
			e.preventDefault();
			this.props.update(this.props.record, this.uri, this.isCreate)
			this.back();
		}
		
		render() {
			if (this.props.loading || !this.props.record) {
				return <Loading />
			}
			const {
				updateField,
				update,
				...remainingProps
			} = this.props;
			return (
				<div>
					{this.renderTitle()}
					<WrappedComponent fieldUpdate={this.handleUpdateField} update={this.handleUpdate} {...remainingProps} />
					{this.renderBackBtn()}
				</div>
			);
		}
		
		renderTitle() {
			const {
				title
			} = this.props;
			return <h1>{title}</h1>;
		}
		
		back() {
			const {
				history
			} = this.props;
			history.goBack();
		}
		
		renderBackBtn() {
			return (
				<Button text="Back" onClick={this.back} />
			);
		}
		
	}

	DetailsComponent.propTypes = {
			load: PropTypes.func.isRequired,
			loading: PropTypes.bool.isRequired,
			history: PropTypes.object.isRequired,
			record: PropTypes.object,
			title: PropTypes.string.isRequired,
			url: PropTypes.string
	};
	
	return DetailsComponent;
}

export default detailsComponent;