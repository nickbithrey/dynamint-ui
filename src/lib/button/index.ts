import { Button, IButton } from './Button';
import { ButtonContainer, IButtonContainer } from './ButtonContainer';

export { 
    Button,
    IButton,
    ButtonContainer,
    IButtonContainer
};