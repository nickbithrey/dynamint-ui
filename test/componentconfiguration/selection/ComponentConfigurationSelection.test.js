import React from 'react';
import { shallow } from 'enzyme';

import Selection from '~/componentconfiguration/selection/ComponentConfigurationSelection';
import { Selection as SelDetails } from 'office-ui-fabric-react/lib/DetailsList';

describe('<ComponentConfigurationSelection /> rendering', () => {

	let props = {
		load: jest.fn(() => []),
		items: [],
		location: { pathname: 'compconfigs'},
		uri: 'uri'
	};
	
	let wrapper;
	
	beforeEach(() => {
		wrapper = shallow( <Selection {...props}/> );
	})
	
	it('matches snapshot', () => {
		expect(wrapper).toMatchSnapshot();
	});
		
	test('resolving selected record from selection details', () => {
		const selDetails = new SelDetails();
		selDetails._items = [{reference: 'ref', uri: 'uri'}];
		selDetails.setIndexSelected(1, true, true);
		const state = wrapper.find('SelectionComponent').prop('buttons')[0].stateFn(selDetails);
		expect(state).toEqual({id: 'ref', uri: 'uri'});
	});
	
	it('will only show edit button when record is selected', () => {
		const selDetails = new SelDetails();
		const editBtn = wrapper.find('SelectionComponent').prop('buttons')[0];
		let condition = editBtn.condition(selDetails);
		expect(condition).toBeFalsy();
		
		selDetails._items = [{reference: 'ref', uri: 'uri'}];
		selDetails.setIndexSelected(1, true, true);
		condition = editBtn.condition(selDetails);
		expect(condition).toBeTruthy();
	});
	
	it('will always show create button', () => {
		const selDetails = new SelDetails();
		const createBtn = wrapper.find('SelectionComponent').prop('buttons')[1];
		let condition = createBtn.condition(selDetails);
		expect(condition).toBeTruthy();
		
		selDetails._items = [{reference: 'ref', uri: 'uri'}];
		selDetails.setIndexSelected(1, true, true);
		condition = createBtn.condition(selDetails);
		expect(condition).toBeTruthy();
	});
	
});