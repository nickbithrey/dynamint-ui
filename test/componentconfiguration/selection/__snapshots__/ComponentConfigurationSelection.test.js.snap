// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ComponentConfigurationSelection /> rendering matches snapshot 1`] = `
<SelectionComponent
  buttons={
    Array [
      SelectionButton {
        "condition": [Function],
        "path": "compconfigs/edit",
        "stateFn": [Function],
        "text": "Edit",
      },
      SelectionButton {
        "condition": [Function],
        "path": "compconfigs/create",
        "stateFn": [Function],
        "text": "Create",
      },
    ]
  }
  columns={
    Array [
      Object {
        "fieldName": "reference",
        "key": "reference",
        "maxWidth": 150,
        "minWidth": 50,
        "name": "Reference",
      },
      Object {
        "fieldName": "componentType",
        "key": "componentType",
        "maxWidth": 50,
        "name": "Type",
      },
      Object {
        "fieldName": "description",
        "key": "description",
        "minWidth": 500,
        "name": "Description",
      },
    ]
  }
  list={Array []}
  loadList={[MockFunction]}
  title="Component Configurations"
/>
`;
